src/js/experiments/snippetFeedback.js

```diff
@@ -0,0 +1,250 @@
+            removeClickListener();
+          }
+        };
+
+        const removeClickListener = () => {
+          document.removeEventListener('click', outsideClickListener);
+        };
+
+        document.addEventListener('click', outsideClickListener);
+      }, 0);
+    }
+  }
+
+  /**
+   * Render the char count and set it's result on the respective element.
+   * */
+  renderCharCount(charCount) {
+    let charCountLimited =
+      charCount > this.maxCharCount ? this.maxCharCount : charCount;
+    const el = `
+        <div class="charCountRow">
+          <span>${charCountLimited} / ${this.maxCharCount}</span>
+        </div>`;
+    this.currCharCountElement.innerHTML = el;
+  }
+
+  // -- Trackers --
+  // These are responsible for sending data based on user input
+  // to amplitude.
+
+  /**
+   * trackYesOrNoButton sends a trackAction to aplitude.
+   * actions are: "docs-snippet-helpful-no"  and "docs-snippet-helpful-yes"
+   * */
+  trackYesOrNoButton(yesOrNoString) {
+    window.AnalyticsClient.trackAction(
+      `docs-snippet-helpful-${yesOrNoString}`,
+      {
+        originatingSnippet: this.snippetElement.textContent,
+        location: window.location.pathname,
+        timeOfButtonClick: Date.now(),
+      },
+    );
+  }
+
+  /**
+   * `trackFormSubmission` invokes trackAction with a payload of the original
+   * snippet, the feedback of the user, and the timeOfFormSubmission.
+   * */
+  trackFormSubmission(formContent) {
+    window.AnalyticsClient.trackAction(`docs-snippet-helpful-form-submission`, {
+      originatingSnippet: this.snippetElement.textContent,
+      feedback: formContent,
+      location: window.location.pathname,
+      wasThisHelpful: this.wasThisHelpful,
+      // for diffing this against the time that the trackYesOrNoButton was clicked
+      // as well as possible the time the copy code button was clicked (TODO add that)
+      timeOfFormSubmission: Date.now(),
+    });
+  }
+
+  // -- Helpers --
+
+  /**
+   * Construct a dom element, setting classes, text content, class, onclick etc.
+   * */
+  makeElement({ kind, text, className, onClick }) {
+    let el = document.createElement(kind);
+    if (text) el.textContent = text;
+    if (className) el.classList.add(className);
+    if (onClick) el.addEventListener('click', onClick);
+
+    return el;
+  }
+}
+
+/**
+ * init queries the dom for all snipppets
+ * TODO(blocked): when that snippet's COPY CODE button is clicked... (waiting for graduation of copy code buttons)
+ *
+ * */
+function init() {
+  // add event listeners to all code snippets on page
+  window.OptimizelyClient.getVariationName({
+    experimentKey: 'dd_docs_snippet_feedback_experiment_test',
+    groupExperimentName: 'q4_fy22_docs_disco_experiment_group_test',
+    experimentContainer: '.hljs',
+  }).then((variation) => {
+    if (variation === 'treatment') {
+      const snippets = document.querySelectorAll('code.hljs');
+      [...snippets].forEach((snippetBlock) => {
```

Comment:
for (let snippetBlock of snippets)

Reply:
or a conventional for loop since index is useful for trackActions

Reply:
or if you want to use the forEach, you could call it directly off the prototype.

main point is the spread feels superfluous.

Reply:
we're dong a lot pretty eagerly when there's no guarantee the user will even copy a snippet to begin with.

I think it would behoove us to consolidate this part code with the highlight js service and consider how we want to do this in a more lazy fashion - whether that's by only initializing this feedback instance when a cursor hovers over a code snippet block OR by using intersection observer API.

Reply:
> I think it would behoove us to consolidate this part code with the highlight js service and consider how we want to do this in a more lazy fashion - whether that's by only initializing this feedback instance when a cursor hovers over a code snippet block OR by using intersection observer API.

So in this case, we'd only initiate the `getVariationName` call when we when we see a code block scroll into view? I feel like the additional code complexity (and having to run a callback on scrolling) could be more complex than an async call that won't block anything anyway?

Reply:
the getVariationName should ideally be called as early as possible since that's a latency issue (and also only needs to run once per page anyways).

What can be lazier is just the scripts to add the input for a given snippet. We can discuss this in a huddle if this is still unclear.

Reply:
Agree with @nramabad. Usually, for this sort of work if you really need the element in the DOM you will only add it once and just show it / move it to the right position once needed.

Reply:
I'm not sure I'm clear on this, so let's pair on this tomorrow after I clean up other issues and fix UI suggestions.